File: Programmers/코딩테스트 고득점 Kit/cpp/42862_체육복.cpp
// 19 20 25 테스트 실패
// 예외 : 7, { 2,4 }, { 1,3 } 정답 : 7
#include<bits/stdc++.h>

#define EL "\n";

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
	// 전체 학생
	vector<int> v(n, 1);
	int answer = n;
	// 체육복을 잃어버린 학생 인덱스값 감소
	for (int i = 0; i < lost.size(); i++) {
		v[lost[i]-1]--;
	}

	// 여분의 체육복을 가진 학생 인덱스값 증가
	for (int i = 0; i < reserve.size(); i++) {
		v[reserve[i]-1]++;
	}

	for (int i = 0; i < v.size(); i++) {
		// i번째 학생이 체육복이 없다면
		if (v[i] == 0) {
			// 앞번호 학생이 여분의 체육복이 있다면
			if (i - 1 >= 0 && v[i - 1] == 2) {
				// 빌려주기
				v[i + 1]--;
				v[i]++;
			}
			// 뒷번호 학생이 여분의 체육복이 있다면
			else if (i + 1 < n && v[i + 1] == 2) {
				// 빌려주기
				v[i - 1]--;
				v[i]++;
			}
			else {
				answer--;
			}
		}
	}

	return answer;
}

void print(int n, vector<int> lost, vector<int> reserve, int answer) {
	int t = solution(n, lost, reserve);
	if (t == answer) {
		cout << "정답" << endl;
	}
	else {
		cout << "틀림" << endl;
	}
}

int main() {
	print(5, { 2, 4 }, { 1, 3, 5 }, 5);
	print(5, { 2, 4 }, { 3 }, 4);
	print(3, { 3 }, { 1 }, 2);
	print(7, { 2, 3, 4 }, { 1, 2, 3, 6 }, 6);
	print(5, { 5 }, { 4 }, 5);
	print(7, { 2,4 }, { 1,3 }, 7);

	return 0;
}

File: Programmers/코딩테스트 고득점 Kit/csharp/42862_체육복.cs
namespace cs
{
    class MainApp
    {
        static void Main()
        {            
            Console.WriteLine(TestCase(5, new int[] { 2, 4 }, new int[] { 1, 3, 5 }, 5));
            Console.WriteLine(TestCase(5, new int[] { 2, 4 }, new int[] { 3 }, 4));
            Console.WriteLine(TestCase(3, new int[] { 3 }, new int[] { 1 }, 2));
        }

        static string TestCase(int n, int[] lost, int[] reserve, int returnValue)
        {
            if(Solution(n, lost, reserve) == returnValue)
                return "성공";

            return "실패";
        }

        static int Solution(int n, int[] lost, int[] reserve)
        {
            // 전체학생에서 체육복 없는 학생 빼기
            int answer = n;

            // 전체 학생이 체육복을 하나씩 가지고 있음
            int[] students = new int[n];
            Array.Fill(students, 1);

            // 체육복을 잃어버린 학생
            for(int i = 0 ; i < lost.Length ; i++)
                students[lost[i] - 1]--;

            // 여벌의 체육복을 챙긴 학생
            for(int i = 0 ; i < reserve.Length ; i++)
                students[reserve[i] - 1]++;

            // 전체 학생 탐색
            for(int i=0 ;i<students.Length ;i++)
            {
                // 체육복을 잃어버린 학생이라면
                if(students[i] == 0)
                {
                    // 앞의 학생에게 빌려보기
                    // 앞의 학생의 인덱스가 0 이상이여야 하고
                    // 앞의 학생의 체육복 개수가 2이상이여야함
                    if(i-1 >= 0 && students[i-1] > 1)
                    {
                        // 앞의 학생 체육복 빌리기
                        students[i - 1]--;
                        students[i]++;
                    }
                    // 뒤의 학생에게 빌려보기
                    // 뒤의 학생의 인덱스가 students.Length보다 작아야하고
                    // 뒤의 학생의 체육복 개수가 2이상이여야함
                    else if(i+1<students.Length && students[i+1] > 1)
                    {
                        // 뒤의 학생 체육복 빌리기
                        students[i + 1]--;
                        students[i]++;
                    }

                    // 앞, 뒤의 학생들에게 빌릴 수 없다면
                    else
                        answer--;
                }
            }
            return answer;
        }
    }
}